crates/astria-sequencer/src/grpc/state_ext.rs

```diff
@@ -272,124 +107,230 @@ pub(crate) trait StateReadExt: StateRead {
-
-        let key = rollup_ids_by_hash_key(&block.block_hash());
-
-        self.nonverifiable_put_raw(
-            key,
-            borsh::to_vec(&RollupIdSeq(rollup_ids))
-                .wrap_err("failed to serialize rollup IDs list")?,
-        );
-
-        let key = sequencer_block_header_by_hash_key(&block.block_hash());
         let SequencerBlockParts {
             block_hash,
             header,
             rollup_transactions,
             rollup_transactions_proof,
             rollup_ids_proof,
         } = block.into_parts();
-        let header = header.into_raw();
-        self.nonverifiable_put_raw(key, header.encode_to_vec());
 
-        for (id, rollup_data) in rollup_transactions {
-            let key = rollup_data_by_hash_and_rollup_id_key(&block_hash, &id);
-            self.nonverifiable_put_raw(key, rollup_data.into_raw().encode_to_vec());
-        }
+        put_block_hash(self, header.height(), block_hash)?;
+        put_rollup_ids(self, &block_hash, rollup_transactions.keys().copied())?;
+        put_block_header(self, &block_hash, header)?;
+        put_rollups_transactions(self, &block_hash, rollup_transactions.into_iter())?;
+        put_rollups_transactions_proof(self, &block_hash, rollup_transactions_proof)?;
+        put_rollup_ids_proof(self, &block_hash, rollup_ids_proof)
+    }
+}
 
-        let key = rollup_transactions_proof_by_hash_key(&block_hash);
-        self.nonverifiable_put_raw(key, rollup_transactions_proof.into_raw().encode_to_vec());
+#[instrument(skip_all)]
+async fn get_sequencer_block_by_hash<S: StateRead + ?Sized>(
+    state: &S,
+    hash: &[u8; 32],
+) -> Result<SequencerBlock> {
+    let header = state
+        .get_sequencer_block_header_by_hash(hash)
+        .await
+        .wrap_err("failed to get sequencer block header by hash")?;
+    let rollup_ids = state
+        .get_rollup_ids_by_block_hash(hash)
+        .await
+        .wrap_err("failed to get rollup ids by block hash")?;
+    let rollup_transactions_proof = state
+        .get_rollup_transactions_proof_by_block_hash(hash)
+        .await
+        .wrap_err("failed to get rollup transactions proof by block hash")?;
+    let rollup_ids_proof = state
+        .get_rollup_ids_proof_by_block_hash(hash)
+        .await
+        .wrap_err("failed to get rollup ids proof by block hash")?;
+
+    #[expect(
+        clippy::default_trait_access,
+        reason = "want to avoid explicitly importing `index_map` crate to sequencer crate"
+    )]
+    let mut parts = SequencerBlockParts {
+        block_hash: *hash,
+        header,
+        rollup_transactions: Default::default(),
+        rollup_transactions_proof,
+        rollup_ids_proof,
+    };
 
-        let key = rollup_ids_proof_by_hash_key(&block_hash);
-        self.nonverifiable_put_raw(key, rollup_ids_proof.into_raw().encode_to_vec());
+    for rollup_id in rollup_ids {
+        let rollup_txs = state.get_rollup_data(hash, &rollup_id).await?;
```

Comment:
1. please provide error context
2. this executes async futures in a loop. We should replace this by a stream. Am fine doing this in a followup, but we should track it.

Reply:
Point 1 done in [dd3bfa6](https://github.com/astriaorg/astria/pull/1492/commits/dd3bfa68494d3d14555a09c0dfb5cbe6228b6f7d).

Point 2 tracked at #1606.